src/System.Diagnostics.Debug/tests/TraceDebugTests.cs

```diff
@@ -0,0 +1,484 @@
+            VerifyLogged(() => Debug.WriteLine("pizza"), expectedIndentTwice + "pizza" + Environment.NewLine);
+
+            // Unindent.
+            Debug.Unindent();
+            Assert.Equal(expectedIndentLevel, Debug.IndentLevel);
+            Debug.Unindent();
+        }
+
+        static void VerifyLogged(Action test, string expectedOutput)
+        {
+            FieldInfo writeCoreHook = typeof(DebugProvider).GetField("s_WriteCore", BindingFlags.Static | BindingFlags.NonPublic);
+
+            // First use our test logger to verify the output
+            var originalWriteCoreHook = writeCoreHook.GetValue(null);
+            writeCoreHook.SetValue(null, new Action<string>(WriteLogger.s_instance.WriteCore));
+
+            try
+            {
+                WriteLogger.s_instance.Clear();
+                test();
+#if DEBUG
+                Assert.Equal(expectedOutput, WriteLogger.s_instance.LoggedOutput);
+#else
+                Assert.Equal(string.Empty, WriteLogger.s_instance.LoggedOutput);
+#endif
+            }
+            finally
+            {
+                writeCoreHook.SetValue(null, originalWriteCoreHook);
+            }
+
+            // Then also use the actual logger for this platform, just to verify
+            // that nothing fails.
+            test();
+        }
+
+        static void VerifyAssert(Action test, params string[] expectedOutputStrings)
+        {
+            FieldInfo writeCoreHook = typeof(DebugProvider).GetField("s_WriteCore", BindingFlags.Static | BindingFlags.NonPublic);
+            s_defaultProvider = Debug.SetProvider(WriteLogger.s_instance);
+
+            var originalWriteCoreHook = writeCoreHook.GetValue(null);
+            writeCoreHook.SetValue(null, new Action<string>(WriteLogger.s_instance.WriteCore));
+
+            try
+            {
+                WriteLogger.s_instance.Clear();
+                test();
+#if DEBUG
+                for (int i = 0; i < expectedOutputStrings.Length; i++)
+                {
+                    Assert.Contains(expectedOutputStrings[i], WriteLogger.s_instance.LoggedOutput);
+                    Assert.Contains(expectedOutputStrings[i], WriteLogger.s_instance.AssertUIOutput);
+                }
+#else
+                Assert.Equal(string.Empty, WriteLogger.s_instance.LoggedOutput);
+                Assert.Equal(string.Empty, WriteLogger.s_instance.AssertUIOutput);
+#endif
+
+            }
+            finally
+            {
+                writeCoreHook.SetValue(null, originalWriteCoreHook);
+                Debug.SetProvider(s_defaultProvider);
+            }
+        }
+
+        private static DebugProvider s_defaultProvider;
+        private class WriteLogger : DebugProvider
+        {
+            public static readonly WriteLogger s_instance = new WriteLogger();
+
+            private WriteLogger() { Assert.Equal(1, Trace.Listeners.Count); } // Activates Wiring of TraceListeners to Debug
```

Comment:
Note:

The new `TraceDebugTests.cs` class contains all `DebugTests` class tests (testing Debug) as well as more specifically testing Trace.
The difference with DebugTests class is line 460:

`private WriteLogger() { Assert.Equal(1, Trace.Listeners.Count); } // Activates Wiring of TraceListeners to Debug`

In other words:
- When TraceDebugTests class is run alone:
It will test Debug and Trace behavior after Trace Listeners are attached with Debug
- When DebugTests class is run alone:
It will test Debug behavior before TraceListeners are attached with Debug.

The two test classes combined show the Debug behavior is the same in both cases.

The combination of both classes helped me refactor Debug and Trace classes more confidently while doing the refactoring and fixing bugs.
They can be removed after the bugfixes and refactoring gets merged in.